Created challenge newsort

SanJSp opened the pull request:
Hi everybody,

I'm looking for early feedback **and help** regarding my contest challenge *newsort*.

It's a fairly simple task, but I think it was a good match between story and task. The students have to find an increasing order in a list of values, but the order of the digits are differently than in our number system. Nothing too special but IMO still a good test. This PR introduces the changes necessary to build this new problem including generator, solution, test cases and notes in `notes.md`.

I'd love to get feedback on:
- the test cases → do you think they cover the problem space well enough?
- the input value sizes → I was “inspired” by the ones @PFischbeck proposed for his O(n) solution (biggest). But I'm insecure if these ranges fit my O(n) challenge well or if I should adapt them.


I need help with:
- my test case `worst_runtime_but_positive.in` → the test case returns "YES" on my system but the generated result is "NO". I was not able to figure out why.
- I'd love to add `sample4`in the latex file but doing this makes my `make` build fail. Do you know why?

Looking forward to your feedback. Thank you for your time and any comments you make! 🙌

Commit:
initial setup of problem with finished executables

Commit:
renamed exercise and adapted test case generator with long time testcase

Commit:
eset unwanted changes

Commit:
added slide notes

Commit:
enhanced generator and removed bugs

Commit:
created problem description

Commit:
reduced problem difficulty

Commit:
reduced problem difficulty from 2 to 1

PFischbeck commented on cp22/hpi-fun2/newsort/problem.tex:
```diff
@@ -0,0 +1,28 @@
+
+Rick and Morty are on a new adventure. This time they landed in a parallel universe and of course, Morty broke Ricks laser gun which is needed ot get back home. To get the money back for the materials for the gun they try to gamble at the stock market. Rick knows that it's a physical law, that the right stocks always go up - and thats where they want to invest. 
+
+You need to figure out if the values of a stock price you receive are increasing, so that Rick and Morty know where to invest. The problem is: In this parallel universe they use the same arabian digits but they don't have the same ordering. In our world we have the ordering $0 < 1 < 2 < 3 < 4 < 5 ...$ but in a parallel universe this ordering could be $6 < 2 < 1 < 4 \dots$ . This ordering does not ...
```
Make sure that it is clear this doesn't mean strictly increasing, but same values are allowed.

PFischbeck commented on cp22/hpi-fun2/newsort/problem.tex:
```diff
@@ -0,0 +1,28 @@
+
+\paragraph*{Input}
+
+The first line contains the number of prices $1\leq n<10^9$ of a stock you have followed by the n prices $0<n_i<10^9$. Last you will receive the digit order of the local universe, wich is a permutation of the digits 0-9.
```
It is not clear from this sentence that this is actually three different lines. Please rephrase :)

I think you mean `n` should be at most `10^5` here. I also think you should change it to be `\leq 10^5`, since this makes it easier to read. Make sure this is consistent with your test cases.

Use `$n$ prices`. Also, for the range of values here, I think it would be easier to not talk about the price being in some range, but rather that it is a digit string with a maximum length of 9.

It is also unclear from the task whether leading zeros are allowed (or whatever a zero is in the new digit system). I think you even have testcases with this.

SanJSp replied:
> Also, for the range of values here, I think it would be easier to not talk about the price being in some range, but rather that it is a digit string with a maximum length of 9.

I'd let it be that way, as this increases the difficulty a bit. This may already push them in the direction to think about a map and so on. But if you insist I can change it :)

PFischbeck commented on cp22/hpi-fun2/newsort/problem.tex:
```diff
@@ -0,0 +1,28 @@
+  \sample{sample1}
+  \sample{sample2}
+  \sample{same_values}
+%  \samlpe{sample4}
```
There is a typo here (`samlpe`), which is why this didn't compile.

SanJSp replied:
🤦‍♂️

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,55 @@
+write_testcase('sample3', 'Sample 3', [47, 40, 99], 2164783059)
+write_testcase('sample4', 'Sample 4', [0, 40, 99], '0164783259')
+write_testcase('wrong_order', 'wrong order', [1, 2], 1234567890)
+write_testcase('same_values', 'numbers are the same', [1, 1], 1234567890)
```
This and the three following testcases have `1234567890` instead of `0123456789`. Is this intended?

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,55 @@
+write_testcase('sample4', 'Sample 4', [0, 40, 99], '0164783259')
+write_testcase('wrong_order', 'wrong order', [1, 2], 1234567890)
+write_testcase('same_values', 'numbers are the same', [1, 1], 1234567890)
+write_testcase('different_lengths', 'values have different lengths', [1, 11111], 1234567890)
```
In this case, both numbers are actually 0. Is this intended?

SanJSp replied:
I think I did not clarify enough, what the number ordering string stands for. I assume you understood it this way: 
num_ordering[0] -> 1 means that in the new universe, the digit ordering replaces our 0 with a 1. 
I meant it this way:
num_ordering[0] -> 1 means that the smallest digit in this universe is 1. Therefore if we had the num_ordering 9876543210 we have 9 as the smallest digit and 0 as the biggest. And thats where I see the problem, you cant really write 0000 as a number :/ 
Do you think its then best to simply say leading zeros are allowed and its simply a complete randomization of the digits? So probably in the way you already understood above, that there is a mapping between index and digit. This could allow 0000 to be different to 00.

SanJSp replied:
I'll try to implement it that way for now :)

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,55 @@
+write_testcase('wrong_order', 'wrong order', [1, 2], 1234567890)
+write_testcase('same_values', 'numbers are the same', [1, 1], 1234567890)
+write_testcase('different_lengths', 'values have different lengths', [1, 11111], 1234567890)
+write_testcase('different_lengths2', 'values have different lengths', [51111, 123333, 15666666], 1234567890)
```
In this case, there are leading zeroes (ones in this case). Is this intended? This is actually not correctly handled by your solution, which doesn't consider leading zeros.

SanJSp replied:
I have fixed it by saying that no leading zeros (or whatever its representation is) exist - I filtered them out in the testcases

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,55 @@
+
+
+MAX_N = 2 * 10**5
+MAX_VAL = 10**9
```
As I suggest in another comment, you could change this to instead be a maximal length (9).

SanJSp replied:
I've let it be that way, as it was easier to generate numbers that way.

PFischbeck replied:
Okay, but for example for `max_input`, it should then be `MAX_VAL - 1`. If you used `MAX_PRICE_LEN`, this would then be `"9" * MAX_PRICE_LEN` :)

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,55 @@
+
+numbersArray.sort()
+write_testcase('worst_runtime_but_positive', 'ascending til the end', numbersArray, '0123456789')
+numbersArray[-1] = 0
```
Make sure it is clear in the task whether 0 is an acceptable value or not.

PFischbeck commented on cp22/hpi-fun2/newsort/executables/solution.cpp:
```diff
@@ -0,0 +1,62 @@
+        if (s1 == s2) {
+            continue;
+        }
+        if(s1.length() > s2.length()){
```
This should actually be `<`, and `>` two lines later. Also, consider the case of leading zeros.

PFischbeck reviewed (commented):
Most of the testcases have the standard digit numbering. You could also try to generate some testcases that have a different numbering. To generate positive testcases in this scenario, you could generate the sorted number sequence, and then replace digits according to the new numbering.

Commit:
pr feedback: fixed leading zeros, enhanced description and added testcasses

SanJSp commented:
Thank you for your feedback! I appreciate it and have already made some changes:
- enhanced testcases with new random ordering
- fixed leading 0 issue I've realized [here](https://github.com/compprog-lecture-tools/fun-contest-2022/pull/1#discussion_r917354514)
- fixed minor errors

PFischbeck commented on cp22/hpi-fun2/newsort/executables/generator.py:
```diff
@@ -0,0 +1,87 @@
+    val_without_leading_zeros = str(val).lstrip(zero_equiv)
+
+    if len(val_without_leading_zeros) == 0:
+        return int(zero_equiv)
+    return int(str(val).lstrip(zero_equiv))
```
In both of these cases, you should not use `int()`, since this would remove leading zeros in our digit system, even though they might not even mean zero in the new system.
So I suggest:

```python
if len(val_without_leading_zeros) == 0:
    return zero_equiv
return val_without_leading_zeros
```

SanJSp replied:
Hi Philipp - I've tried this now but it renders the sort() functions below (line 72) useless and therefore the random test cases that should be positive end up negative. Do you maybe have an alternative idea for this?

PFischbeck replied:
Ah, I see. The problem at that point is that we are sorting strings instead of ints, but in this particular case there actually can't be any leading zeros (because `zero_equiv = "0"`). Two ideas:

1. When sorting in this case, use `numbersArray.sort(key=int)`, which treats everything as ints when comparing.
2. Or: rewrite the whole approach to generating the random test cases, where you still have control over whether it's ordered or not. In particular: You generate a sequence of ints, those are the values in the normal numbering system. You can then rearrange them to be sorted if you want (for a positive testcase). Then, you choose the digit permutation. Then, for every int, you replace the digits according to the permutation, and this gives you your testcase input. This is basically what you did for the `worst_runtime_random_positive` testcase, but generalized. In particular, you don't have to care about leading zeros anymore, since you replace every digit with the correct other digit. I can help out with this or write it, if you want :)

SanJSp replied:
Hi again,
I've been trying around with your first idea but unfortunately the test cases still end up as false negatives. I would really appreciate it, if you could implement your second idea. Do you think this would be possible? I'd really be thankful for this.

PFischbeck replied:
No worries! I implemented the second idea now, giving us additional positive + negative testcases for random numbers and random orderings.

PFischbeck commented on cp22/hpi-fun2/newsort/problem.tex:
```diff
@@ -0,0 +1,28 @@
+\begin{samples}
+  \sample{sample1}
+  \sample{sample2}
+  \sample{same_values}
```
You can't know this yet: All testcases starting with `sample` will be available for download in the judge, all others are not. Thus, you should name all sample testcases starting with `sample`.

PFischbeck committed:
Fix typos in problem description

PFischbeck committed:
Add philipp solution

PFischbeck commented:
Thanks for the improvements! See my small comments. In addition, if you want, you can add further random testcases in the style of the one you added, which generate random numbers and a random ordering, and then replace the digits accordingly. This is optional however.

If you can give me push/edit rights for this PR, I can add my own Python solution.

SanJSp commented:
Thanks for the response 🚀  I appreciate your feedback 😊  and will take care of it by the end of next week 💯  
I've invited you as collaborator to my fork, I hope thats what you meant with:

> If you can give me push/edit rights for this PR, I can add my own Python solution.

If not, I will try to find an alternative way.

Commit:
changed sample names

PFischbeck committed:
Fix max_input testcase

PFischbeck committed:
Add more random mostly-ordered testcases
